File: sort/c_implement/select_sort.c
//
// Created by kaku on 2019/10/7.
//

#include "array.h"

void select_sort(int *arr, int array_size) {
    int max_val;
    int max_idx;
    int temp;

    for (int i = 0; i < array_size; i++) {
        max_val = arr[0];
        max_idx = 0;
        for (int j = 1; j < array_size - i; j++) {
            if (max_val < arr[j]) {
                max_idx = j;
                max_val = arr[j];
            }
        }

        temp = arr[max_idx];
        arr[max_idx] = arr[array_size - i - 1];
        arr[array_size - i - 1] = temp;
    }
}

File: sort/python_implement/select_sort.py
#!/usr/bin/env python3
# coding: utf-8
#
# author: kaku
# date: 19/10/08

# GitHub:
#
#   https://github.com/kakukosaku
#
# © 2019-2022 Kaku Kosaku All Rights Reserved

from typing import List, NoReturn
from .array import swap


def select_sort(arr: List[int], array_size: int) -> NoReturn:
    """Select Sort

    Notes:
        1. In Python arguments pass by reference to mutable variables, needn't return arr.
        2. Pass array size to function `even in Python can get array(list) len on runtime`.
        3. Replace for loop with while since Python for loop is not friendly to use index.

        4. Find the Right position; Move for space; Put it in.

    """
    i = 0
    while i < array_size:
        min_idx = i
        j = i + 1
        while j < array_size:
            # from small to big
            if arr[j] < arr[min_idx]:
                min_idx = j
            j += 1

        swap(arr, i, min_idx)
        i += 1
